### PR TITLE
Add workloads for assertion; make it ready for continuous test

### PR DESCRIPTION
**Pull request by ekexium**
Add a workload T2 for assertion, and make other improvements in error handling, logging, etc.

### CODE REVIEW

File: src/workload.rs [hunk 1]
```diff
@@ -144,32 +174,108 @@ impl Workload for DoubleInsertion {
...
+                "github.com/pingcap/tidb/table/tables/corruptMutations",
+            )
+            .await?;
+        }
+
+        Ok(())
+    }
+}
+
+// 2 txns. The first writes corrupted data and the second reads it. Check if Assertion can detect it.
```
Comment (MyonKeminta) [inline src/workload.rs]: The second txn should be a write too (and your code is write...) Assertion is expected to take effect when trying to write correct data on existing corrupted data or write corrupted data on existing correct data.

File: src/workload.rs [hunk 2]
```diff
@@ -144,32 +174,108 @@ impl Workload for DoubleInsertion {
...
+        send!(conn, query(create_statement.as_str())).expect("don't let create statement fail");
+        let duration = start.elapsed();
+        CREATE_TABLE_DURAION_MS.fetch_add(duration.as_millis() as u64, Ordering::SeqCst);
+        info!(log, "{}", create_statement);
+
+        for injection in AVAILABLE_INJECTIONS {
+            info!(log, "{} ready to go!", injection);
+
+            // NOTE: "1*" here, otherwise an index mutation is missing for each row insertion, thus cannot be detected.
+            enable_failpoint(
```
Comment (MyonKeminta) [inline src/workload.rs:326]: Actually, I think you can consider inject it on either the `insert` or the `update`. I think assertion should possibly take effect on both cases.

File: src/table.rs [hunk 3]
```diff
@@ -38,7 +37,7 @@ impl ToString for ColumnType {
 }
 
 #[derive(Clone, Debug, Hash, PartialEq, Eq)]
-enum Datum {
+pub enum Datum {
     Int(i64),
     String(String),
```
Comment (MyonKeminta) [inline src/table.rs:42]: Maybe `varchar(n)` and `char(n)` can be tested respectively

File: src/workload.rs [hunk 4]
```diff
@@ -144,32 +174,108 @@ impl Workload for DoubleInsertion {
...
+                table.cols[1].name,
+                row.cols[1].to_string()
+            );
+
+            let res = async {
+                send!(conn, query(insertion.as_str()))?;
+                send!(conn, "commit")?;
+                send!(conn, "begin optimistic")?;
+                send!(conn, query(deletion.as_str()))?;
+                send!(conn, query(update.as_str()))?;
```
Comment (MyonKeminta) [inline src/workload.rs]: Does this mean delete and then update? What's the purpose here?

File: src/workload.rs (hunk 1 above)
Comment (ekexium) [reply src/workload.rs]: Right

File: src/workload.rs (hunk 2 above)
Comment (ekexium) [reply src/workload.rs:326]: OK let's make it another workload.

File: src/workload.rs (hunk 4 above)
Comment (ekexium) [reply src/workload.rs]: It's a mistake. Should be a delete after update.
I added the update stmt to see if it catches some cases of corruption that the deletion failed to detect.

File: src/table.rs (hunk 3 above)
Comment (ekexium) [reply src/table.rs:42]: We can. I don't quite understand why it is needed. Is there a big difference between them?

File: src/workload.rs [hunk 5]
```diff
@@ -115,62 +158,289 @@ impl Workload for SingleInsertion {
...
+
+            // NOTE: "1*" here, otherwise an index mutation is missing for each row insertion, thus cannot be detected.
+            enable_failpoint(
+                &log,
+                client,
+                "github.com/pingcap/tidb/table/tables/corruptMutations",
+                format!("1*return(\"{}\")", injection),
+            )
+            .await?;
+            send!(log, conn, "begin optimistic")?;
```
Comment (MyonKeminta) [inline src/workload.rs]: Is there any reason to begin it here? It seems putting this statement near following statements at L272 makes the code easier to read

File: src/workload.rs [hunk 6]
```diff
@@ -115,62 +158,289 @@ impl Workload for SingleInsertion {
...
+        let start = Instant::now();
+        let duration = start.elapsed();
+        CREATE_TABLE_DURAION_MS.fetch_add(duration.as_millis() as u64, Ordering::SeqCst);
+
+        for injection in AVAILABLE_INJECTIONS {
+            send!(log, conn, drop_statement.as_str()).expect("don't let drop statement fail");
+            send!(log, conn, create_statement.as_str()).expect("don't let create statement fail");
+            info!(log, "{} ready to go!", injection);
+
+            // NOTE: "1*" here, otherwise an index mutation is missing for each row insertion, thus cannot be detected.
```
Comment (MyonKeminta) [inline src/workload.rs]: It seems the comment is not in the correct place.

File: src/workload.rs [hunk 7]
```diff
@@ -115,62 +158,289 @@ impl Workload for SingleInsertion {
...
+                table.cols[0].name,
+                row.cols[0].next().to_string(),
+                table.cols[1].name,
+                row.cols[1].to_string()
+            );
+            let deletion = format!(
+                "DELETE FROM {} WHERE {} = {}",
+                table.name,
+                table.cols[0].name,
+                row.cols[0].to_string()
```
Comment (MyonKeminta) [inline src/workload.rs]: Is it expected to delete the row by the old value of cols[0], which seems should not exist? Consider describe your purpose in comments

**Review by MyonKeminta — COMMENTED** [review]
Rest LGTM

File: src/workload.rs (hunk 5 above)
Comment (ekexium) [reply src/workload.rs]: Makes sense

File: src/workload.rs (hunk 7 above)
Comment (ekexium) [reply src/workload.rs]: It might be intentional when I wrote it but yeah deleting using new values is more realisitc, I will change it.

**Pull request merged**